vermouth/ffinput.py

```diff
@@ -51,196 +55,275 @@
-            start += 1
-    return tokens
-
-
-def _substitute_macros(line, macros):
-    r"""
-    Substitute macros by their content.
-
-    A macro starts with a '$' and ends with one amongst ' ${}\n\t"'.
-
-    Parameters
-    ----------
-    line: str
-        The line to fix.
-    macros: dict[str, str]
-        Keys are macro names, values are the replacement content.
+class FFDirector(SectionLineParser):
+    COMMENT_CHAR = ';'
+    interactions_natoms = {
+        'bonds': 2,
+        'angles': 3,
+        'dihedrals': 4,
+        'impropers': 4,
+        'constraints': 2,
+        'virtual_sites2': 3,
+        'pairs': 2,
+    }
 
-    Returns
-    -------
-    str
-    """
-    start = None
-    while True:  # stops when start < 0
-        start = line.find('$', start)
-        if start < 0:
-            break
-        for end, char in enumerate(line[start + 1:], start=start + 1):
-            if char in ' \t\n{}$"':
-                break
-        else: # no break
-            end += 1
-        macro_name = line[start + 1:end]
-        macro_value = macros[macro_name]
-        line = line[:start] + macro_value + line[end:]
-        end = start + len(macro_value)
-    return line
+    def __init__(self, force_field):
+        super().__init__()
+        self.force_field = force_field
+        self.current_block = None
+        self.current_link = None
+        self.current_modification = None
+        self.blocks = collections.OrderedDict()
+        self.links = []
+        self.modifications = []
+
+        self.header_actions = {
+            ('moleculetype', ): self._new_block,
+            ('link', ): self._new_link,
+            ('modification', ): self._new_modification,
+        }
+
+    def parse_header(self, line, lineno=0):
+        """
+        Parses a section header with line number `lineno`. Sets :attr:`section`
+        when applicable. Does not check whether `line` is a valid section
+        header.
+
+        Parameters
+        ----------
+        line: str
+        lineno: str
+
+        Returns
+        -------
+        object
+            The result of calling :meth:`finalize_section`, which is called
+            if a section ends.
+
+        Raises
+        ------
+        KeyError
+            If the section header is unknown.
+        """
+        prev_section = self.section
+        
+        ended = []
+        section = self.section + [line.strip('[ ]').casefold()]
+        if tuple(section[-1:]) in self.METH_DICT:
+            self.section = section[-1:]
+        else:
+            wildcard_section = tuple(section[:-1] + ['*'])
+            while (tuple(section) not in self.METH_DICT
+                   and wildcard_section not in self.METH_DICT
+                   and len(section) > 1):
+                ended.append(section.pop(-2))  # [a, b, c, d] -> [a, b, d]
+                wildcard_section = tuple(section[:-1] + ['*'])
+            self.section = section
```

Comment:
This code in combination with 152 fails to actually do what it is supposed to do. Instead of dealing with wildcard sections of the type [ALA]['atoms'] it actually only deals with sections of type [moleculetype] [*], which seems a workaround to not enumerate all bonds angles etc. in METH_DICT. In fact it fails for reading the .rtp file in the universal force-field. Also, this is not at all caught by any of the tests.

Reply:
My bad, the read_ff parser is not supposed to read .rtp files. But obviously the rtp file parser is also not a subclass of any of the base parsers in parser_utils.py.